cleanup group spread code and minor modifications

LukasSchaefer opened the pull request.

LukasSchaefer committed:
cleanup group spread code and minor modifications

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -21,25 +22,23 @@ def make_world(self, groups, num_colors=None, reward_per_group=True, randomise_a
+        self.num_colors = len(self.groups) if num_colors is None else num_colors
+        self.obs_onehot_colors = obs_onehot_colors
 
         if not self.randomise_all_colors:
```
Previously colors are randomly generated with `np.random.random(3)` but this can lead to very close/ overlapping colors. I agree this is unlikely but evenly spaced out colors might be easier to distinguish programmatically.

LukasSchaefer committed:
minor change of randomised colors

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -21,25 +22,23 @@ def make_world(self, groups, num_colors=None, reward_per_group=True, randomise_a
-            # generate color per group
-            self.colors = [np.random.random(3) for _ in range(self.num_colors)]
+            # generate color per group evenly spaced out if not randomised
+            self.colors = [np.array([i / (self.num_colors - 1)] * 3) for i in range(self.num_colors)]
```
Previously colors are randomly generated with np.random.random(3) but this can lead to very close/ overlapping colors. I agree this is unlikely but evenly spaced out colors might be easier to distinguish programmatically.

mhairidunion replied:
This now forces R, G and B values to be the same for each colour which seems quite limiting. It also means all episodes will render using a few different shades of grey (making the images difficult for humans). The original MPE tasks seem to hard code the colours to use so we could use the same colours as those tasks I guess but we lose some flexibility then.

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -48,44 +47,37 @@ def make_world(self, groups, num_colors=None, reward_per_group=True, randomise_a
-        # random properties for agents
+            # TODO: do nicer to avoid overlap here?
+            # randomly generate new colors for each episode
+            self.colors = [np.random.random(3) for _ in range(self.num_colors)]
```
If new colors should be created every episode then still keep random colors (would like to change it to avoid overlapping but afaik we don't even use this feature atm?)

mhairidunion replied:
Not sure I see the point here? There should be one colour for each group. The num_colors parameter is used to specify how many colours are available to chose from, which in the case of randomise_all_colors should be all possible colours. I think this is only required if we change the colour generator to use fixed colours/interval as suggested above, otherwise I think it works. Agree that maybe the parameters need new names as we've added new functionality it may have got confusing.

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -48,44 +47,37 @@ def make_world(self, groups, num_colors=None, reward_per_group=True, randomise_a
+            self.colors = [np.random.random(3) for _ in range(self.num_colors)]
+
+        # assign one of the available colors to a group
+        color_idxs = np.random.choice(self.num_colors, len(self.groups), replace=False)
```
Colors indices should always be picked for each group with indices being between (0, `self.num_colors - 1`). Previously, when randomised colors are used, these were between (0, `len(self.groups) - 1`) which is not the same if `num_colors` argument is provided to the environment! E.g. for `num_colors = 3` when there are only 2 groups, this would have avoided to use the 3rd color ever if randomised colors are used. This probably wouldn't have any meaningful effect since colors in that case are randomised anyway, but seems like odd logic and as far as I can tell there is no reason to make this different depending on whether randomised colors are used?

mhairidunion replied:
See comment above - I don't think it makes sense to use num_colours with randomise_all_colors, they are conflicting. If randomised colours are used then the colour is select from all possible colours at the start of episode.

LukasSchaefer replied:
Now simplified to just get new color values in the random case (handling of how these are selected all happens in the color config definition.

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -48,44 +47,37 @@ def make_world(self, groups, num_colors=None, reward_per_group=True, randomise_a
+        for group_id, land_id in zip(self.group_indices, landmarks_ids):
+            world.landmarks[land_id].color = self.colors[color_idxs[group_id]]
+            world.landmarks[land_id].color_idx = color_idxs[group_id]
+            world.landmarks[land_id].group = group_id
```
No need to differentiate this depending on whether the observations are randomly shuffled or not? If they are randomly shuffled, then now they are shuffled twice (assignment of groups to landmarks here is random + again for obs construction) and if not randomly shuffled obs-wise then they are still shuffled here as before.

mhairidunion replied:
Actually now with the changes they are shuffled twice when `shuffle_obs_per_agent=True` since they are shuffled here and again when agent order is assigned. Previously they were only shuffled here if `shuffle_obs_per_agent=False` since they won't get shuffled again later. Although I don't think it really makes much difference and I agree this is much tidier now anyway.

LukasSchaefer replied:
Yes, exactly. Previously, this ensured that everything is only shuffled once whereas now it can be shuffled twice but that shouldn't affect anything (twice uniform shuffling should be same as once) and code becomes simpler to read,

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -48,44 +47,37 @@ def make_world(self, groups, num_colors=None, reward_per_group=True, randomise_a
+                agent.agent_obs_order = np.random.choice(self.num_agents, self.num_agents, replace=False)
+            else:
+                agent.landmark_obs_order = np.arange(self.num_landmarks)
+                agent.agent_obs_order = np.arange(self.num_agents)
```
Now always setting this variable, just depending on the shuffling value and setting it to the arange arrays to have the default behaviour if `shuffle_obs_per_agent=False`.

LukasSchaefer replied:
(Just for cleaner code in obs construction later)

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -150,34 +142,26 @@ def global_reward(self, world):
     def observation(self, agent, world):
+        # own position, velocity, color
+        color = [np.eye(self.num_colors)[agent.color_idx]] if self.obs_onehot_colors else [agent.color]
+        entity_features = [agent.state.p_pos] + [agent.state.p_vel] + color
```
This substitutes the `[agent.state.p_vel] + [agent.state.p_pos] + [np.eye(self.num_colors)[agent.group]]` in the last line of this function before (wanted to make these 3 blocks as separate code).

Also, here I think there are 2 actually meaningful bugs to do with colors of agents:

1. Previously, the color of the agent itself was extracted as `np.eye(self.num_colors)[agent.group]` but I think this should be `np.eye(self.num_colors)[agent.color_idx]` as used now for one-hot colors. At the beginning of each episode, `color_idxs` are randomised (shuffled for each group), so the index access previously was to the wrong color!

2. The previous code always gave the color of the agent as the one-hot array. However, previously the colors of the landmarks were provided as real color values and not one-hots whenever `randomise_all_colors=True` in which case it was impossible for agents to identify which landmark matches their color.

mhairidunion replied:
Yes these are both bugs, I think the first one probably explains the performance issues with 3 agents!

LukasSchaefer replied:
Further simplified color obs generation --> now all from color config defined above

raulsteleac replied:
Great finds, especially 1., will start some runs to check the shuffling problem after we merge this!

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -150,34 +142,26 @@ def global_reward(self, world):
-                entity_pos_color.append(world.landmarks[i].color)
-            else:
-                # one hot encoding
-                entity_pos_color.append(np.eye(self.num_colors)[world.landmarks[i].color_idx])
```
This code conflates one-hot encoding of colors and `randomise_all_colors=True` we use real-valued colors, if `randomise_all_colors=False` then we use one-hot colors for landmarks.

This is imo quite unintuitive as the flag does not suggest this is happening and I also don't see why we would conflate those two effects in this way (isn't it reasonable to use one-hot colors in agent observations even if we use randomised colors, and also to use real-valued colors in observations even if we use consistent colors).

To address this, I use a new `obs_onehot_colors` flag which is true by default and can determine whether all color values in observations (own, other agents, landmarks) are given as onehot (True; default) or real-valued arrays (False).

mhairidunion replied:
I'm fine with a new flag here, I think it makes it clearer. However, I would add the one-hot encoding when randomising all colours doesn't make sense - that would be a very big vector to allow for all colours! So maybe we want to catch that and raise an exception.

LukasSchaefer replied:
Thanks for catching this. Now that's not possible anymore (no config for this weird case of random colors with onehot encoding)

LukasSchaefer committed:
update group spread code -- cleanup color settings
- avoid "unallowed" combinations of arguments through enum and
  ValueError
- make used colors more distinguishable (not use all shades of grey)
- cleanup code -- separate color value generation and color obs values
  generation in enum

LukasSchaefer committed:
fix small error

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -1,91 +1,155 @@
+    np.array([1.00, 0.00, 0.00]),
+    np.array([0.00, 1.00, 0.00]),
+    np.array([0.00, 0.00, 1.00]),
+]
```
To avoid having all grey-scale colors if fixed set of colors are used, I hand-wrote this list of possible colors. Main downside is we are now limited to max 10 colors but it's easy to extend if there would be a need.

raulsteleac replied:
Looks good to me for now, unlikely that we will have more than 10 groups but I'll keep it in mind if there's a way for a more automated approach

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -1,91 +1,155 @@
+class ColorConfig(Enum):
+    CONTINUOUS_FIXED = 0    # used fixed set of colors (matched by `num_colors`) which agents observe as continuous vectors
+    ONEHOT_FIXED = 1        # used fixed set of colors (matched by `num_colors`) which agents observe as one-hot vectors
+    CONTINUOUS_RANDOM = 2   # used random colors which agents observe as continuous vectors (not using `num_colors`!)
```
Define all general options of colors supported here

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -1,91 +1,155 @@
+        elif label.lower().replace(" ", "_") in ("continuousrandom", "continuous_random", "random"):
+            return ColorConfig.CONTINUOUS_RANDOM
+        else:
+            raise NotImplementedError(f"Unknown color config label '{label}'")
```
Simplify generation of color configurations from strings

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -1,91 +1,155 @@
+            return [np.random.random(3) for _ in range(num_colors)]
+        elif self == ColorConfig.ONEHOT_FIXED or self == ColorConfig.CONTINUOUS_FIXED:
+            # generate fixed set of colors using only individual color channels 
+            return COLORS[:num_colors]
```
Use just first colors from the pre-defined fixed set of colors whenever using standard fixed colors.

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -1,91 +1,155 @@
+        elif self == ColorConfig.ONEHOT_FIXED:
+            return np.eye(len(colors))[color_idx]
+        else:
+            raise NotImplementedError(f"Unknown color config '{self.name}'")
```
Use color value in observations for random colors and the `continuous_fixed` option, but one-hot encoded for the onehot option

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -1,91 +1,155 @@
-            # generate color per group
-            self.colors = [np.random.random(3) for _ in range(self.num_colors)]
+        if self.color_config == ColorConfig.CONTINUOUS_RANDOM and num_colors is not None:
+            raise ValueError("Color config is set to CONTINUOUS_RANDOM but `num_colors` is explicitly set. `num_colors` value is not used!")
```
Catch this edge case that was previously weird where `num_colors` are provided but also random colors are used.

raulsteleac replied:
I wouldn't stop the execution for this but maybe just add a warning and override the value instead. 

If we would use **num_colors** for CONTINOUS_RANDOM as well it wouldn't break the run but get more or less the same result. Changes the logic from sampling from e.g 5 random colors instead of 2 random colors for 2 groups (np.random.choice(5, 2, reuse=False)), but for the model it's all the same.

LukasSchaefer replied:
Good point, changed from exception to warning.

LukasSchaefer committed:
update default arguments of group spread and default color ordering

LukasSchaefer committed:
update scenarios code: set default arguments only in main "group_spread.py"
(allow overwriting/ passthrough with env_args given to gym.make)

LukasSchaefer commented on pettingzoo/mpe/group_spread_1p_1p_1p_v0.py:
```diff
@@ -9,7 +9,7 @@ def __init__(self, seed=None, local_ratio=0.5, max_frames=50, **env_args):
         ), "local_ratio is a proportion. Must be between 0 and 1."
         scenario = Scenario()
-        world = scenario.make_world(groups=[1,1,1], num_colors=None, reward_per_group=True, randomise_all_colors=False, shuffle_obs_per_agent=True)
+        world = scenario.make_world(groups=[1, 1, 1], **env_args)
```
Only set group argument in here (other consistent defaults are once defined in `group_spread.py`), and allow `env_args` to be given through to environment.

raulsteleac replied:
Much cleaner and it will show in the config :+1:

LukasSchaefer commented on pettingzoo/mpe/group_spread_3p_v0.py:
```diff
@@ -1,5 +1,5 @@
 from ._mpe_utils.simple_env import SimpleEnv, make_env
-from .scenarios.group_spread_one_hot import Scenario
+from .scenarios.group_spread import Scenario
```
Seemed to be outdated code since `group_spread_one_hot.py` does not exist for me?

raulsteleac replied:
I removed that file but as it was my old implementation (before we merged everything in group_spread) and missed this config. Good find

LukasSchaefer commented on pettingzoo/mpe/scenarios/group_spread.py:
```diff
@@ -1,91 +1,155 @@
 
 class Scenario(BaseScenario):
-    def make_world(self, groups, num_colors=None, reward_per_group=True, randomise_all_colors=False, shuffle_obs_per_agent=False):
+    def make_world(self, groups, reward_per_group=True, color_config="onehot_fixed", num_colors=None, shuffle_obs_per_agent=True):
```
Make sure those are the default config we want to have!

raulsteleac reviewed (commented):
Looks overall great to me, thanks for cleaning up the code!

LukasSchaefer committed:
change exception to warning when num colours set for random colours
